### ISSUE TITLE
Enhance Model Spec to Distinguish Between Base Model and LoRA Adaptor Components

### ISSUE THREAD

**Issue by CormickKneey** [modelpack/model-spec#60]
**Description:**
The current spec for CloudNativeAI/model-spec defines a generic Model grouping weight layers, config, and metadata. However, in the context of parameter-efficient fine-tuning (e.g., LoRA), it’s essential to distinguish the base model and adapter components explicitly. This separation enables more flexible deployment, reusable base model distribution, and smaller, focused adapters.

---

**Motivation / Use Case:**

* Users often start from a large pretrained base model (e.g., LLaMA 13B) and fine-tune lightweight LoRA adapters on domain-specific data.
* They want to package:

  * `base_model`: the full pretrained checkpoint (e.g., `.bin` or safetensor)
  * `lora_adaptor`: small delta weights only (typically <300 MB)
* Use cases:

  1. **Modularity:** Distribute base model once; share multiple adapters across domains.
  2. **Optimization:** Deploy only base+adapter as separate OCI layers—avoids repackaging whole model for each task.
  3. **Reuse:** Host base model in EC2/Harbor; publish adapter packages independently.
  4. **Efficient inference pipelines:** Orchestrators (e.g., Kubernetes) can mount base and adapter layers dynamically.

---

**Proposed Extension to Spec:**

* Add metadata labels, e.g.,:

  ```json
  {
    "org.cnai.model.role": "base" | "lora-adapter",
    "org.cnai.model.parent_digest": "<digest of base model>"
  }
  ```

---

**Benefits:**

| Advantage                   | Description                                                                                       |
| --------------------------- | ------------------------------------------------------------------------------------------------- |
| Reduced storage & bandwidth | Tag and reuse base model only once                                                                |
| Clear adapter metadata      | Facilitates tooling support for loading Merge/Overlay logic                                       |
| OCI-friendly separation     | Aligns with OCI artifact layering semantics  |

---

**Questions & Open Points:**

* Should `ModelConfig` include a field like `"adapterRole": "lora"`?
* How to represent the relationships between lora adapter and base model?

**Comment by aftersnow** [comment]
The LoRA weights and the base model are 1 to 1 correlated. Would it be more appropriate to package them together?  Or are you looking to distinguish the LoRA weights from the base model weights?

**Comment by gaius-qi** [comment]
If the `subject` is deleted in harbor, both OCI artifacts will be deleted.

**Comment by amisevsk** [comment]
> The LoRA weights and the base model are 1 to 1 correlated. Would it be more appropriate to package them together? Or are you looking to distinguish the LoRA weights from the base model weights?
> 
>> Actually, packaging the LoRA weights together with the base model isn't feasible, as in real-world scenarios the LoRA adaptor is highly dynamic — with potentially multiple new versions being released daily. In contrast, the base model remains relatively static. Therefore, the two should be linked in a clearly visible and modular way, rather than being bundled together.

Technically, the OCI format somewhat invalidates this reason; OCI is basically designed to allow layering in a small, frequently updated set of changes (e.g. a server binary) on top of a large, static layer (e.g. a secured runner base image). 

If you package both together, as long as you store the original model weights in a separate layer it will only ever be pushed/pulled once, and will be deduplicated in OCI stores.
